packages/react-server/src/ReactFlightServer.js

```diff
@@ -136,20 +129,17 @@ function attemptResolveElement(element: React$Element<any>): ReactModel {
   } else if (typeof type === 'string') {
     // This is a host element. E.g. HTML.
     return [REACT_ELEMENT_TYPE, type, element.key, element.props];
-  } else if (
-    type === REACT_FRAGMENT_TYPE ||
-    type === REACT_STRICT_MODE_TYPE ||
-    type === REACT_PROFILER_TYPE ||
-    type === REACT_SCOPE_TYPE ||
-    type === REACT_DEBUG_TRACING_MODE_TYPE ||
-    type === REACT_LEGACY_HIDDEN_TYPE ||
-    type === REACT_OFFSCREEN_TYPE ||
-    // TODO: These are temporary shims
-    // and we'll want a different behavior.
-    type === REACT_SUSPENSE_TYPE ||
-    type === REACT_SUSPENSE_LIST_TYPE
-  ) {
-    return element.props.children;
+  } else if (typeof type === 'symbol') {
+    if (type === REACT_FRAGMENT_TYPE) {
+      // For key-less fragments, we add a small optimization to avoid serializing
+      // it as a wrapper.
+      // TODO: If a key is specified, we should propagate its key to any children.
+      // Same as if a server component has a key.
```

Comment:
This actually reminded me.. What's the plan for different types? Do those also get encoded somehow when we get to encoding keys? Or does the type change trick not work for Server Components?

Reply:
Yea, it doesn't work for server components but that was already the plan for function components too as part of the "inlining optimization" or "call through".  At least stateless ones. The reason it needs to kill state is because the state doesn't "line up" with a different type. A stateless doesn't need to.

Reply:
Oh right. So it doesn't matter until you run into a Client one, at which point you have the types on the client anyway.

I guess I still find this a bit counter-intuitive in terms of guarantees. It allows for some fun patterns though. Like SCs could use recursion and then have a terminal CC, and it could still match up on refetch despite different SC depth?

Reply:
Sure. You could also just call a recursive function. I guess you could use it as a resuming hack.